File: app/src/main/java/com/jetsetslow/omniterm/data/AppDatabase.kt
package com.jetsetslow.omniterm.data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase

@Database(
    entities = [
        ServerEntity::class,
        MetricHistoryEntity::class,
        SshKeyEntity::class,
        CredentialProfileEntity::class,
        AlertRuleEntity::class,
        ActiveAlertEntity::class,
        AlertHistoryEntity::class,
        QuickScriptEntity::class,
        WolTargetEntity::class,
        NetworkShareEntity::class,
        AppSettingEntity::class,
        PersistentSessionEntity::class,
        PortForwardEntity::class,
        StackRegistryEntity::class
    ],
    // NOTE: bump this whenever any @Entity schema changes, and add a real Migration for the bump.
    // Schemas are exported to app/schemas/ (committed) so migrations can be written and tested
    // against the exact prior shape. Versions ≤7 predate schema export (several v5 builds shipped
    // with differing schemas), so upgrades from those still fall back to a destructive wipe — but
    // from v8 on, a version bump without a Migration must fail loudly instead of deleting data.
    version = 20,
    exportSchema = true
)
abstract class AppDatabase : RoomDatabase() {
    abstract fun serverDao(): ServerDao
    abstract fun metricHistoryDao(): MetricHistoryDao
    abstract fun sshKeyDao(): SshKeyDao
    abstract fun credentialProfileDao(): CredentialProfileDao
    abstract fun alertRuleDao(): AlertRuleDao
    abstract fun activeAlertDao(): ActiveAlertDao
    abstract fun alertHistoryDao(): AlertHistoryDao
    abstract fun quickScriptDao(): QuickScriptDao
    abstract fun wolTargetDao(): WolTargetDao
    abstract fun networkShareDao(): NetworkShareDao
    abstract fun appSettingDao(): AppSettingDao
    abstract fun persistentSessionDao(): PersistentSessionDao
    abstract fun portForwardDao(): PortForwardDao
    abstract fun stackRegistryDao(): StackRegistryDao

    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        private val MIGRATION_8_9 = object : Migration(8, 9) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE servers ADD COLUMN proxyKeyAlias TEXT")
            }
        }

        // The backup-jobs feature was removed before it ever shipped a UI; drop its table.
        private val MIGRATION_9_10 = object : Migration(9, 10) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("DROP TABLE IF EXISTS backup_jobs")
            }
        }

        // Persistent (tmux-backed) sessions: per-server opt-in flag.
        private val MIGRATION_10_11 = object : Migration(10, 11) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE servers ADD COLUMN persistentSession INTEGER NOT NULL DEFAULT 0")
            }
        }

        // Track live tmux sessions so they can be re-offered after an app restart.
        private val MIGRATION_11_12 = object : Migration(11, 12) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL(
                    "CREATE TABLE IF NOT EXISTS persistent_sessions (" +
                        "tmuxName TEXT NOT NULL PRIMARY KEY, " +
                        "serverId INTEGER NOT NULL, " +
                        "serverName TEXT NOT NULL, " +
                        "createdAt INTEGER NOT NULL)"
                )
            }
        }

        // WoL targets gain an optional host IP, used to ping for live online status.
        private val MIGRATION_12_13 = object : Migration(12, 13) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE wol_targets ADD COLUMN ipAddress TEXT NOT NULL DEFAULT ''")
            }
        }

        // Scripts and alert rules gain a free-text notes/comment field for documentation.
        private val MIGRATION_13_14 = object : Migration(13, 14) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE quick_scripts ADD COLUMN notes TEXT NOT NULL DEFAULT ''")
                db.execSQL("ALTER TABLE alert_rules ADD COLUMN notes TEXT NOT NULL DEFAULT ''")
            }
        }

        // Saved LAN/network share profiles for SMB/FTP/SFTP/NFS/WebDAV discovery and access metadata.
        private val MIGRATION_14_15 = object : Migration(14, 15) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE credential_profiles ADD COLUMN groupName TEXT NOT NULL DEFAULT 'General'")
                db.execSQL(
                    "CREATE TABLE IF NOT EXISTS network_shares (" +
                        "id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, " +
                        "name TEXT NOT NULL, " +
                        "protocol TEXT NOT NULL, " +
                        "address TEXT NOT NULL, " +
                        "port INTEGER NOT NULL, " +
                        "sharePath TEXT NOT NULL, " +
                        "workgroup TEXT NOT NULL, " +
                        "username TEXT NOT NULL, " +
                        "password TEXT NOT NULL, " +
                        "authProfileId INTEGER, " +
                        "anonymous INTEGER NOT NULL, " +
                        "notes TEXT NOT NULL, " +
                        "lastChecked INTEGER NOT NULL, " +
                        "lastStatus TEXT NOT NULL)"
                )
            }
        }

        // WebDAV shares gain an explicit TLS flag; backfill from the old port heuristic (443/8443
        // were treated as https) so existing shares keep connecting exactly as before.
        private val MIGRATION_15_16 = object : Migration(15, 16) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE network_shares ADD COLUMN useHttps INTEGER NOT NULL DEFAULT 0")
                db.execSQL("UPDATE network_shares SET useHttps = 1 WHERE UPPER(protocol) = 'WEBDAV' AND port IN (443, 8443)")
            }
        }

        // Per-server SSH agent forwarding (ssh -A) opt-in + saved port-forward tunnels.
        private val MIGRATION_16_17 = object : Migration(16, 17) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE servers ADD COLUMN agentForwarding INTEGER NOT NULL DEFAULT 0")
                db.execSQL(
                    "CREATE TABLE IF NOT EXISTS port_forwards (" +
                        "id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, " +
                        "serverId INTEGER NOT NULL, " +
                        "name TEXT NOT NULL, " +
                        "kind TEXT NOT NULL DEFAULT 'local', " +
                        "bindHost TEXT NOT NULL DEFAULT '127.0.0.1', " +
                        "bindPort INTEGER NOT NULL, " +
                        "destHost TEXT NOT NULL DEFAULT '', " +
                        "destPort INTEGER NOT NULL DEFAULT 0, " +
                        "autoStart INTEGER NOT NULL DEFAULT 0)"
                )
            }
        }

        // App-side registry of compose stacks, so a stack downed via `compose down` (containers
        // and networks removed — the daemon keeps no record) can still be listed and brought up.
        private val MIGRATION_17_18 = object : Migration(17, 18) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL(
                    "CREATE TABLE IF NOT EXISTS stack_registry (" +
                        "id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, " +
                        "serverId INTEGER NOT NULL, " +
                        "runtime TEXT NOT NULL, " +
                        "project TEXT NOT NULL, " +
                        "workingDir TEXT NOT NULL, " +
                        "configFiles TEXT NOT NULL, " +
                        "lastSeenAt INTEGER NOT NULL)"
                )
                db.execSQL(
                    "CREATE UNIQUE INDEX IF NOT EXISTS index_stack_registry_serverId_runtime_project " +
                        "ON stack_registry (serverId, runtime, project)"
                )
            }
        }

        // One rule can have one live incident per concrete host. Keep the newest legacy row before
        // enforcing that identity; older builds could race manual and periodic telemetry probes.
        private val MIGRATION_18_19 = object : Migration(18, 19) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL(
                    "DELETE FROM active_alerts WHERE id NOT IN (" +
                        "SELECT MAX(id) FROM active_alerts GROUP BY ruleId, serverId)"
                )
                db.execSQL(
                    "CREATE UNIQUE INDEX IF NOT EXISTS index_active_alerts_ruleId_serverId " +
                        "ON active_alerts (ruleId, serverId)"
                )
            }
        }

        // Preset rows get a stable presetKey so the "default presets" toggles can remove exactly
        // what they seeded even after the user edits one. Back-stamping is deliberately gated by
        // the corresponding setting: a matching name/category or fleet-wide metric is not proof
        // that a row belongs to OmniTerm when that preset family was disabled.
        private val MIGRATION_19_20 = object : Migration(19, 20) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE quick_scripts ADD COLUMN presetKey TEXT")
                db.execSQL("ALTER TABLE alert_rules ADD COLUMN presetKey TEXT")
                // "Backgrounded since" for saved tmux sessions. Existing rows have never been
                // observed being backgrounded, so seed them from createdAt rather than "now",
                // which would falsely show every restored session as just-backgrounded.
                db.execSQL("ALTER TABLE persistent_sessions ADD COLUMN backgroundedAt INTEGER NOT NULL DEFAULT 0")
                db.execSQL("UPDATE persistent_sessions SET backgroundedAt = createdAt WHERE backgroundedAt = 0")

                // Fleet presets used to be on by default without persisting a setting. Preserve
                // that state only when a recognisable legacy fleet row exists. A stored false
                // value wins because INSERT OR IGNORE never overwrites the user's choice.
                db.execSQL(
                    "INSERT OR IGNORE INTO app_settings (`key`, value) " +
                        "SELECT 'fleet_presets', 'true' WHERE EXISTS (" +
                        "SELECT 1 FROM quick_scripts WHERE category = 'Fleet' AND name IN (" +
                        LEGACY_SCRIPT_PRESET_KEYS
                            .filter { it.familySetting == "fleet_presets" }
                            .joinToString(",") { "'${it.name.replace("'", "''")}'" } +
                        "))"
                )
                for (preset in LEGACY_SCRIPT_PRESET_KEYS) {
                    db.execSQL(
                        "UPDATE quick_scripts SET presetKey = ? " +
                            "WHERE presetKey IS NULL AND name = ? AND category = ? " +
                            "AND EXISTS (SELECT 1 FROM app_settings WHERE `key` = ? AND value = 'true')",
                        arrayOf(preset.key, preset.name, preset.category, preset.familySetting),
                    )
                }
                for (preset in LEGACY_RULE_PRESETS) {
                    db.execSQL(
                        "UPDATE alert_rules SET presetKey = ? " +
                            "WHERE presetKey IS NULL AND serverId = 0 AND metricName = ? " +
                            "AND mountPoint = '/' AND thresholdValue = ? AND severity = ? " +
                            "AND triggerWindow = '5m' AND enabled = 1 AND notes = '' " +
                            "AND EXISTS (SELECT 1 FROM app_settings " +
                            "WHERE `key` = 'alert_presets' AND value = 'true')",
                        arrayOf<Any>(
                            preset.key,
                            preset.metric,
                            preset.threshold,
                            preset.severity,
                        ),
                    )
                }
            }
        }

        private data class LegacyScriptPreset(
            val key: String,
            val name: String,
            val category: String,
            val familySetting: String,
        )

        /** Seeded identities for rows created before the presetKey column existed. */
        private val LEGACY_SCRIPT_PRESET_KEYS = listOf(
            LegacyScriptPreset("fleet.cpu", "CPU/RAM", "Fleet", "fleet_presets"),
            LegacyScriptPreset("fleet.disk", "Disk", "Fleet", "fleet_presets"),
            LegacyScriptPreset("fleet.processes", "Processes", "Fleet", "fleet_presets"),
            LegacyScriptPreset("fleet.services", "Failed services", "Fleet", "fleet_presets"),
            LegacyScriptPreset("fleet.syslog", "Syslog errors", "Fleet", "fleet_presets"),
            LegacyScriptPreset("fleet.containers", "Containers", "Fleet", "fleet_presets"),
            LegacyScriptPreset("fleet.ports", "Listening ports", "Fleet", "fleet_presets"),
            LegacyScriptPreset("fleet.kernel", "Kernel", "Fleet", "fleet_presets"),
            LegacyScriptPreset("homelab.pve_vms", "PVE: list VMs", "Proxmox", "homelab_presets"),
            LegacyScriptPreset("homelab.pve_containers", "PVE: list containers", "Proxmox", "homelab_presets"),
            LegacyScriptPreset("homelab.pve_cluster", "PVE: cluster status", "Proxmox", "homelab_presets"),
            LegacyScriptPreset("homelab.pve_storage", "PVE: storage status", "Proxmox", "homelab_presets"),
            LegacyScriptPreset("homelab.pve_start_vm", "PVE: start VM <id>", "Proxmox", "homelab_presets"),
            LegacyScriptPreset("homelab.pve_stop_vm", "PVE: stop VM <id>", "Proxmox", "homelab_presets"),
            LegacyScriptPreset("homelab.casaos_status", "CasaOS: status", "CasaOS", "homelab_presets"),
            LegacyScriptPreset("homelab.casaos_restart", "CasaOS: restart", "CasaOS", "homelab_presets"),
            LegacyScriptPreset("homelab.casaos_version", "CasaOS: version", "CasaOS", "homelab_presets"),
            LegacyScriptPreset("homelab.ha_info", "HA: info", "Home Assistant", "homelab_presets"),
            LegacyScriptPreset("homelab.ha_core_logs", "HA: core logs", "Home Assistant", "homelab_presets"),
            LegacyScriptPreset("homelab.ha_restart_core", "HA: restart core", "Home Assistant", "homelab_presets"),
            LegacyScriptPreset("homelab.ha_supervisor_logs", "HA: supervisor logs", "Home Assistant", "homelab_presets"),
            LegacyScriptPreset("homelab.temperature", "Temperature", "Linux", "homelab_presets"),
            LegacyScriptPreset("homelab.updates", "Updates available", "Homelab", "homelab_presets"),
            LegacyScriptPreset("homelab.reboot_required", "Reboot required?", "Homelab", "homelab_presets"),
            LegacyScriptPreset("homelab.top_cpu", "Top 10 by CPU", "Homelab", "homelab_presets"),
            LegacyScriptPreset("homelab.docker_stats", "Docker stats", "Homelab", "homelab_presets"),
            LegacyScriptPreset("homelab.disk_usage", "Disk usage", "Homelab", "homelab_presets"),
        )

        private data class LegacyRulePreset(
            val key: String,
            val metric: String,
            val threshold: Float,
            val severity: String,
        )

        /** Exact pristine alert identities predating presetKey. */
        private val LEGACY_RULE_PRESETS = listOf(
            LegacyRulePreset("alert.cpu", "CPU Usage", 90f, "CRITICAL"),
            LegacyRulePreset("alert.memory", "Memory Usage", 90f, "CRITICAL"),
            LegacyRulePreset("alert.disk", "Disk Usage", 90f, "WARNING"),
            LegacyRulePreset("alert.latency", "Latency", 250f, "WARNING"),
        )

        /** Complete non-destructive migration chain for builders and schema regression tests. */
        internal val ALL_MIGRATIONS: Array<Migration> = arrayOf(
            MIGRATION_8_9,
            MIGRATION_9_10,
            MIGRATION_10_11,
            MIGRATION_11_12,
            MIGRATION_12_13,
            MIGRATION_13_14,
            MIGRATION_14_15,
            MIGRATION_15_16,
            MIGRATION_16_17,
            MIGRATION_17_18,
            MIGRATION_18_19,
            MIGRATION_19_20,
        )

        fun getDatabase(context: Context): AppDatabase {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    "omniterm_database"
                )
                // Destructive only for the un-exported legacy versions; never for v8+.
                .fallbackToDestructiveMigrationFrom(dropAllTables = true, 1, 2, 3, 4, 5, 6, 7)
                .addMigrations(*ALL_MIGRATIONS)
                .build()
                INSTANCE = instance
                instance
            }
        }
    }
}

File: flutter_app/lib/data/app_database.dart
import 'dart:io';

import 'package:drift/drift.dart';
import 'package:drift/native.dart';
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'package:sqlite3/sqlite3.dart';

import 'dao/alerts_dao.dart';
import 'dao/app_data_dao.dart';
import 'dao/server_dao.dart';
import 'legacy_presets.dart';
import 'tables.dart';

part 'app_database.g.dart';

/// The Room database file name. Must not change: on Android the Flutter app opens the *same file*
/// the shipped Kotlin app created, so a user updating from the native build keeps every host, key,
/// script and alert.
const kDatabaseFileName = 'omniterm_database';

/// The schema version the shipped Android app reached (`AppDatabase.kt`, `version = 22`).
///
/// This number is load-bearing for data continuity. Android's `SQLiteOpenHelper` — which Room is
/// built on — records the schema version in `PRAGMA user_version`, and that is the same pragma
/// Drift reads. So a database already migrated to Room v22 is seen by Drift as "already at version
/// 22" and no migration runs: the existing rows simply open.
///
/// (The README's mention of "schema v18" is stale; the exported schemas in
/// `app/schemas/com.jetsetslow.omniterm.data.AppDatabase/` run to 22.json.)
const kRoomSchemaVersion = 22;

@DriftDatabase(
  tables: [
    Servers,
    MetricHistory,
    SshKeys,
    CredentialProfiles,
    AlertRules,
    ActiveAlerts,
    AlertHistory,
    QuickScripts,
    WolTargets,
    NetworkShares,
    AppSettings,
    PersistentSessions,
    PortForwards,
    StackRegistry,
  ],
  daos: [ServerDao, AlertsDao, AppDataDao],
)
class AppDatabase extends _$AppDatabase {
  AppDatabase([QueryExecutor? executor]) : super(executor ?? _open());

  /// In-memory instance for tests.
  AppDatabase.memory() : super(NativeDatabase.memory());

  @override
  int get schemaVersion => kRoomSchemaVersion;

  @override
  MigrationStrategy get migration => MigrationStrategy(
    onCreate: (m) async {
      await m.createAll();
    },
    onUpgrade: (m, from, to) async {
      // Room's own chain, ported step for step. It normally never runs: a device coming from
      // the native app is already at 22, and a fresh install starts there. It matters only for
      // a user who installs the Flutter build directly over an older native build without
      // taking the intermediate Android update.
      //
      // Versions <= 7 predate schema export (several v5 builds shipped with differing
      // schemas), so Room fell back to a destructive wipe for those and so do we. From v8 on
      // every step is non-destructive.
      if (from <= 7) {
        await _recreateEverything(m);
        return;
      }
      await _runRoomMigrations(m, from);
    },
    beforeOpen: (details) async {
      await customStatement('PRAGMA foreign_keys = ON');
    },
  );

  Future<void> _recreateEverything(Migrator m) async {
    for (final table in allTables) {
      await m.deleteTable(table.actualTableName);
    }
    await m.createAll();
  }

  /// Replays `AppDatabase.ALL_MIGRATIONS` from `data/AppDatabase.kt`.
  Future<void> _runRoomMigrations(Migrator m, int from) async {
    Future<void> step(int target, Future<void> Function() body) async {
      if (from < target) await body();
    }

    await step(9, () async {
      await customStatement('ALTER TABLE servers ADD COLUMN proxyKeyAlias TEXT');
    });

    // The backup-jobs feature was removed before it ever shipped a UI; drop its table.
    await step(10, () async {
      await customStatement('DROP TABLE IF EXISTS backup_jobs');
    });

    // Persistent (tmux-backed) sessions: per-server opt-in flag.
    await step(11, () async {
      await customStatement(
        'ALTER TABLE servers ADD COLUMN persistentSession INTEGER NOT NULL DEFAULT 0',
      );
    });

    // Track live tmux sessions so they can be re-offered after an app restart.
    await step(12, () async {
      await customStatement(
        'CREATE TABLE IF NOT EXISTS persistent_sessions ('
        'tmuxName TEXT NOT NULL PRIMARY KEY, '
        'serverId INTEGER NOT NULL, '
        'serverName TEXT NOT NULL, '
        'createdAt INTEGER NOT NULL)',
      );
    });

    // WoL targets gain an optional host IP, used to ping for live online status.
    await step(13, () async {
      await customStatement(
        "ALTER TABLE wol_targets ADD COLUMN ipAddress TEXT NOT NULL DEFAULT ''",
      );
    });

    // Scripts and alert rules gain a free-text notes/comment field for documentation.
    await step(14, () async {
      await customStatement("ALTER TABLE quick_scripts ADD COLUMN notes TEXT NOT NULL DEFAULT ''");
      await customStatement("ALTER TABLE alert_rules ADD COLUMN notes TEXT NOT NULL DEFAULT ''");
    });

    // Saved LAN/network share profiles for SMB/FTP/SFTP/NFS/WebDAV discovery and access metadata.
    await step(15, () async {
      await customStatement(
        "ALTER TABLE credential_profiles ADD COLUMN groupName TEXT NOT NULL DEFAULT 'General'",
      );
      await customStatement(
        'CREATE TABLE IF NOT EXISTS network_shares ('
        'id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, '
        'name TEXT NOT NULL, '
        'protocol TEXT NOT NULL, '
        'address TEXT NOT NULL, '
        'port INTEGER NOT NULL, '
        'sharePath TEXT NOT NULL, '
        'workgroup TEXT NOT NULL, '
        'username TEXT NOT NULL, '
        'password TEXT NOT NULL, '
        'authProfileId INTEGER, '
        'anonymous INTEGER NOT NULL, '
        'notes TEXT NOT NULL, '
        'lastChecked INTEGER NOT NULL, '
        'lastStatus TEXT NOT NULL)',
      );
    });

    // WebDAV shares gain an explicit TLS flag; backfill from the old port heuristic (443/8443 were
    // treated as https) so existing shares keep connecting exactly as before.
    await step(16, () async {
      await customStatement(
        'ALTER TABLE network_shares ADD COLUMN useHttps INTEGER NOT NULL DEFAULT 0',
      );
      await customStatement(
        "UPDATE network_shares SET useHttps = 1 "
        "WHERE UPPER(protocol) = 'WEBDAV' AND port IN (443, 8443)",
      );
    });

    // Per-server SSH agent forwarding (ssh -A) opt-in + saved port-forward tunnels.
    await step(17, () async {
      await customStatement(
        'ALTER TABLE servers ADD COLUMN agentForwarding INTEGER NOT NULL DEFAULT 0',
      );
      await customStatement(
        'CREATE TABLE IF NOT EXISTS port_forwards ('
        'id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, '
        'serverId INTEGER NOT NULL, '
        'name TEXT NOT NULL, '
        "kind TEXT NOT NULL DEFAULT 'local', "
        "bindHost TEXT NOT NULL DEFAULT '127.0.0.1', "
        'bindPort INTEGER NOT NULL, '
        "destHost TEXT NOT NULL DEFAULT '', "
        'destPort INTEGER NOT NULL DEFAULT 0, '
        'autoStart INTEGER NOT NULL DEFAULT 0)',
      );
    });

    // App-side registry of compose stacks, so a stack downed via `compose down` (containers and
    // networks removed — the daemon keeps no record) can still be listed and brought up.
    await step(18, () async {
      await customStatement(
        'CREATE TABLE IF NOT EXISTS stack_registry ('
        'id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, '
        'serverId INTEGER NOT NULL, '
        'runtime TEXT NOT NULL, '
        'project TEXT NOT NULL, '
        'workingDir TEXT NOT NULL, '
        'configFiles TEXT NOT NULL, '
        'lastSeenAt INTEGER NOT NULL)',
      );
      await customStatement(
        'CREATE UNIQUE INDEX IF NOT EXISTS index_stack_registry_serverId_runtime_project '
        'ON stack_registry (serverId, runtime, project)',
      );
    });

    // One rule can have one live incident per concrete host. Keep the newest legacy row before
    // enforcing that identity; older builds could race manual and periodic telemetry probes.
    await step(19, () async {
      await customStatement(
        'DELETE FROM active_alerts WHERE id NOT IN ('
        'SELECT MAX(id) FROM active_alerts GROUP BY ruleId, serverId)',
      );
      await customStatement(
        'CREATE UNIQUE INDEX IF NOT EXISTS index_active_alerts_ruleId_serverId '
        'ON active_alerts (ruleId, serverId)',
      );
    });

    await step(20, _migrate19To20);

    // Temperature is optional because many VMs, containers, and remote OSes expose no sensor.
    await step(21, () async {
      await customStatement('ALTER TABLE metric_history ADD COLUMN cpuTemperatureC REAL');
    });

    // metric_history was never indexed, so every widget/monitoring read grouped and re-probed it
    // with full scans whose cost tracks retained history (7 days by default), not fleet size. Past
    // the widget's 8s load budget that turned into a stuck "Saving…" during widget configuration.
    // Measured at 150k rows: 469s -> 0.008s.
    await step(22, () async {
      await customStatement(
        'CREATE INDEX IF NOT EXISTS index_metric_history_serverId_timestamp '
        'ON metric_history (serverId, timestamp)',
      );
    });
  }

  /// Preset rows get a stable presetKey so the "default presets" toggles can remove exactly what
  /// they seeded even after the user edits one.
  ///
  /// Back-stamping is deliberately gated by the corresponding setting: a matching name/category or
  /// fleet-wide metric is not proof that a row belongs to OmniTerm when that preset family was
  /// disabled.
  Future<void> _migrate19To20() async {
    await customStatement('ALTER TABLE quick_scripts ADD COLUMN presetKey TEXT');
    await customStatement('ALTER TABLE alert_rules ADD COLUMN presetKey TEXT');

    // "Backgrounded since" for saved tmux sessions. Existing rows have never been observed being
    // backgrounded, so seed them from createdAt rather than "now", which would falsely show every
    // restored session as just-backgrounded.
    await customStatement(
      'ALTER TABLE persistent_sessions ADD COLUMN backgroundedAt INTEGER NOT NULL DEFAULT 0',
    );
    await customStatement(
      'UPDATE persistent_sessions SET backgroundedAt = createdAt WHERE backgroundedAt = 0',
    );

    // Fleet presets used to be on by default without persisting a setting. Preserve that state only
    // when a recognisable legacy fleet row exists. A stored false value wins because INSERT OR
    // IGNORE never overwrites the user's choice.
    final fleetNames = kLegacyScriptPresets
        .where((preset) => preset.familySetting == 'fleet_presets')
        .map((preset) => "'${preset.name.replaceAll("'", "''")}'")
        .join(',');
    await customStatement(
      'INSERT OR IGNORE INTO app_settings (`key`, value) '
      "SELECT 'fleet_presets', 'true' WHERE EXISTS ("
      "SELECT 1 FROM quick_scripts WHERE category = 'Fleet' AND name IN ($fleetNames))",
    );

    for (final preset in kLegacyScriptPresets) {
      await customStatement(
        'UPDATE quick_scripts SET presetKey = ? '
        'WHERE presetKey IS NULL AND name = ? AND category = ? '
        "AND EXISTS (SELECT 1 FROM app_settings WHERE `key` = ? AND value = 'true')",
        [preset.key, preset.name, preset.category, preset.familySetting],
      );
    }

    for (final preset in kLegacyRulePresets) {
      await customStatement(
        'UPDATE alert_rules SET presetKey = ? '
        'WHERE presetKey IS NULL AND serverId = 0 AND metricName = ? '
        "AND mountPoint = '/' AND thresholdValue = ? AND severity = ? "
        "AND triggerWindow = '5m' AND enabled = 1 AND notes = '' "
        'AND EXISTS (SELECT 1 FROM app_settings '
        "WHERE `key` = 'alert_presets' AND value = 'true')",
        [preset.key, preset.metric, preset.threshold, preset.severity],
      );
    }
  }
}

/// Opens the database at the location the native Android app used, so an updating user's data is
/// found rather than silently replaced by an empty database.
///
/// Room resolves its file through `Context.getDatabasePath(name)`, i.e. `<app data>/databases/`.
/// `path_provider` has no API for that directory, but it does expose the application *support*
/// directory (`<app data>/files` on Android), whose parent is the app data root — so the Room
/// directory is derivable from it. On iOS there is no legacy database to inherit; the same layout
/// is used purely for consistency.
QueryExecutor _open() {
  return LazyDatabase(() async {
    final support = await getApplicationSupportDirectory();
    final dir = Directory(p.join(p.dirname(support.path), 'databases'));
    if (!dir.existsSync()) dir.createSync(recursive: true);

    if (Platform.isAndroid) {
      // Android's default temp location is not writable from the sqlite3 native library, so point
      // it at the app's own cache directory before any statement that spills to disk.
      //
      // The old `applyWorkaroundToOpenSqlite3OnOldAndroidVersions()` call is deliberately absent:
      // `sqlite3_flutter_libs` is a no-op stub as of 0.6.0+eol and is obsolete once `sqlite3` 3.x
      // bundles the library itself, so the obsolete compatibility package has been dropped.
      sqlite3.tempDirectory = (await getTemporaryDirectory()).path;
    }

    return NativeDatabase.createInBackground(File(p.join(dir.path, kDatabaseFileName)));
  });
}
